---
title: "Analyse Motif Gains at Disordered Cytosolic Tails of Transmembrane Proteins using Humsavar variants"
output:
  html_document:
    code_folding: hide
    style: spacelab
    toc: yes
    toc_float: yes
date: "`r date()`"
---

```{r setup, echo = FALSE, message=FALSE, warning=FALSE}
workdir <- '/data/local/buyar/collaborations/katrina/analysis_motif_gains_TM_proteins'
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir = workdir)
library(slimR)
library(rtracklayer)
library(Biostrings)
library(data.table)
library(ggplot2)
library(stringi)
library(parallel)
library(biomaRt)
library(ggrepel)
```


# Introduction

This vignette explains how to screen the proteome for gains of short linear motifs via disease-causing or polymorphic single amino-acid variations. 

- Variation data: Humsavar database
- Short linear motif annotations: ELM database
- Disorder predictions using IUPred tool. 
- Protein sequence and feature annotations from Uniprot database. 

We make use of the slimR package along with other R/Bioconductor packages to show how to screen the 
transmembrane proteins for gains/losses of short linear motifs. Moreover, this code shows the results that were obtained and published in the paper: https://www.sciencedirect.com/science/article/pii/S0092867418310353

# Prepare datasets
```{r getData}
# download uniprot fasta and import  
fastaFile <- file.path(workdir, slimR::getUniprotData(format = 'fasta'))
fasta <- Biostrings::readAAStringSet(fastaFile)
#save additional info as metadata
l <- as.list(names(fasta))
names(l) <- sub("^sp\\|(.+?)\\|.*", "\\1", names(fasta))
metadata(fasta) <- l
#process fasta headers to keep only uniprot accessions
names(fasta) <- names(l)

# download uniprot gff annotations and import
gffFile <- file.path(workdir, slimR::getUniprotData(format = 'gff'))
gff <- rtracklayer::import.gff(gffFile)

#download slim regular expression and add an extra motif (diLeu motif)
slim_regex <- slimR::getElmClasses()
slim_regex <- c(slim_regex, list('LIG_diLeu_1' = '.LL.'))

# run iupred to get disorder predictions 
iupred <- slimR::runIUPred(iupredPath = '/home/buyar/tools/iupred', 
                           fastaFile = fastaFile, 
                           nodeN = 10)
#process to keep only uniprot accessions
names(iupred) <- sub("^sp\\|(.+?)\\|.*", "\\1", names(iupred))

#download single amino acid variants 
humsavar <- slimR::getHumSavar()

#add disorder score field to each variant
humsavar$iupredScore <- pbapply::pbapply(as.data.table(humsavar), 1, function(x) {
  uni <- x[['seqnames']]
  pos <- as.numeric(x[['start']])
  score <- iupred[[uni]][pos]
  return(score)
})

#find out if variants overlap any known domains 
humsavar$overlapsDomain <- FALSE
humsavar[unique(queryHits(findOverlaps(humsavar, gff[gff$type == 'Domain',]))),]$overlapsDomain <- TRUE

```

# Find motif gains/losses due to humsavar variants

To predict motif gains/losses in protein sequences, we need to have a 
table of variants (with columns: 

1. uniprot accession 
2. position of variant
3. wild-type amino acid
4. mutant amino acid 

a list of regular expressions each representing a different class
of short linear motifs. 

We also need to look for motif gains/losses in disordered regions of the proteome. Also, we should avoid 
regions of the proteome which are known to harbour protein domains. Point mutations in such ordered regions are most likely disruptive due to their effects on the protein structure, rather than the effect on protein-protein interactions mediated by short linear motifs. 

```{r}
#Get humsavar variants in disordered regions and not overlapping
#any protein domains (as annotated in uniprot),
#remove variants that are 'Unclassified'. 
variants <- as.data.table(humsavar[humsavar$iupredScore >= 0.4 & 
                       humsavar$overlapsDomain == FALSE & 
                       humsavar$variant %in% c('Disease', 'Polymorphism'),])
```

Some humsavar variants may be represented multiple times 
because they may be associated to multiple diseases, 
but when looking for motif changes caused by each variant, 
we want to count variants only once. 

```{r}
#subset variants for the necessary columns and get a unique list of single amino acid variants 
variants <- subset(variants, select = c('seqnames', 'start', 'wtAA', 'mutAA', 'variant'))
colnames(variants) <- c('uniprotAccession', 'pos', 'wtAA', 'mutAA', 'variant_type')
variants <- unique(variants)

#assign a unique label for each variant
variants$variantID <- paste(variants$uniprotAccession, 
                                 variants$wtAA,
                                 variants$pos,
                                 variants$mutAA, sep = ':')

```

Now we are ready to look for motif gains/losses due to single amino acid variants. 
```{r motifChanges}
#look for motifs in proteins with at least one variant annotation
#looking for motif changes for 15574 variants in 6391 protein sequences takes about 10 minutes 
motifChanges <- slimR::findMotifChangesMulti(
  sequences = fasta[names(fasta) %in% unique(variants$uniprotAccession)], 
  variants = variants, 
  nodeN = 10)

motifChanges$names <- metadata(fasta)[motifChanges$uniprotAccession]
```

Let's make some summary plots for the consequences of the screened variants 
```{r}
#merge motifChanges table with the variants table
motifChanges <- merge(motifChanges, variants, by = c('uniprotAccession', 'pos', 'wtAA', 'mutAA'))
```

# Figure 7B and 7C from Meyer et al. 

Define a function to filter variants by overlap with regions 
```{r}

getVariantsByRegion <- function(variants, regions) {
  variants <- makeGRangesFromDataFrame(df = variants, 
                                       seqnames.field = 'uniprotAccession', 
                                       start.field = 'pos',
                                       end.field = 'pos',
                                       ignore.strand = T 
                                       )
  return(unique(queryHits(findOverlaps(variants, regions))))
}

compareMotifChangesByVariantType <- function(vars, motifChanges, changeType = 'gained') {

  s <- motifChanges[variantID %in% vars$variantID & 
                      change == changeType & 
                      variant_type %in% c('Polymorphism', 'Disease'),
             length(variantID), 
             by = c('SLiM', 'variant_type', 'change')]
  
  ds <- dcast.data.table(s, SLiM ~ variant_type, value.var = 'V1')
  ds[is.na(ds)] <- 0
  
  t <- table(vars$variant_type)
  
  ds$DiseaseBG <- t[['Disease']]
  ds$PolymorphismBG <- t[['Polymorphism']]
  
  ds <- cbind(ds, do.call(rbind, apply(ds[,2:5], 1, function(x) {
    p <- fisher.test(matrix(x, nrow = 2))$p.val
    fc <- (x[1]+1)/(x[3]+1) * (x[4]+1)/(x[2]+1)
    return(data.frame('pval' = p, 'foldChange' = fc))
  })))
  
  return(ds)
}

```


```{r}
gff.dt <- as.data.table(gff)
#cytosolic regions of transmembrane proteins
ctm <- makeGRangesFromDataFrame(df = gff.dt[type == 'Topological domain' & Note == 'Cytoplasmic',])
#extracellular regions of transmembrane proteins
etm <- makeGRangesFromDataFrame(df = gff.dt[type == 'Topological domain' & Note == 'Extracellular',])
#list of transmembrane proteins with at least one annotatoin of Cytoplasmic or Extracellular regions
tmProteins <- union(as.character(seqnames(ctm)), as.character(seqnames(etm)))
```


## Reproduce Figure 7B
```{r}
mc <- motifChanges[SLiM == 'LIG_diLeu_1']

l <- list('all_proteins' = compareMotifChangesByVariantType(vars = variants, 
                                  motifChanges = mc, 
                                  changeType = 'gained'),
     'transmembrane_proteins' = compareMotifChangesByVariantType(
       vars =  variants[variants$uniprotAccession %in% tmProteins,], 
                                  motifChanges = mc, 
                                  changeType = 'gained'),
     'transmembrane_extracellular' = compareMotifChangesByVariantType(
       vars = variants[getVariantsByRegion(variants, etm),], 
                                  motifChanges = mc, 
                                  changeType = 'gained'),
     'transmembrane_cytoplasmic' = compareMotifChangesByVariantType(
       vars = variants[getVariantsByRegion(variants, ctm),], 
                                  motifChanges = mc, 
                                  changeType = 'gained'))
df <- do.call(rbind, lapply(names(l), function(r) {
  df <- l[[r]]
  df$region <- r
  return(df)
}))

ggplot2::ggplot(data = df,
                  aes(x = region, y = log2(foldChange))) +
    geom_bar(aes(fill = ifelse(pval < 0.05, 'red', 'black')), position = 'dodge', stat = 'identity') +
  geom_text(aes(label = paste('p =',round(pval, 3)), 
                y = log2(foldChange)/2), 
            color = 'white', size = 6) + 
  scale_fill_identity()
    
```

## Reproduce Figure 7C

```{r Figure_7C}
#cytoplasmic regions of transmembrane proteins
vars = variants[getVariantsByRegion(variants, ctm),]
dt <- compareMotifChangesByVariantType(vars = vars, motifChanges = motifChanges, changeType = 'gained')
ggplot(dt, aes(x = log2(foldChange), y = -log10(pval))) + 
  geom_point(aes(color = ifelse(pval < 0.05, 'red', 'black')), size = 5, alpha = 0.4) + 
  geom_label_repel(aes(label = ifelse(pval < 0.05, SLiM, ''))) + 
  scale_color_identity()
```

